File: govern/data-security/ranger-2.3.0/security-admin/src/main/webapp/scripts/models/RangerPolicy.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

 
define(function(require){
	'use strict';	

	var RangerPolicyBase	= require('model_bases/RangerPolicyBase');
	var XAUtils		= require('utils/XAUtils');
	var XAEnums		= require('utils/XAEnums');
	var localization= require('utils/XALangSupport');

	var RangerPolicy = RangerPolicyBase.extend(
	/** @lends RangerPolicy.prototype */
	{
		/**
		 * RangerPolicy initialize method
		 * @augments RangerPolicyBase
		 * @constructs
		 */
		initialize: function() {
			this.modelName = 'RangerPolicy';
			this.bindErrorEvents();
		},
		/**
		 * @function schema
		 * This method is meant to be used by UI,
		 * by default we will remove the unrequired attributes from serverSchema
		 */

		schemaBase : function(){
			var attrs = _.omit(this.serverSchema, 'id', 'createDate', 'updateDate', "version",
					"permList", "status", "updatedBy", "isSystem");

			_.each(attrs, function(o){
				o.type = 'Hidden';
			});

			// Overwrite your schema definition here
			return _.extend(attrs,{
				name : {
                    type		: 'TextFieldWithIcon',
					title		: 'Policy Name *',
					validators	: ['required'],
					editorAttrs 	:{ maxlength: 255},
                                        errorMsg    : localization.tt("validationMessages.policyNameValidationMsg")
				},
				description : {
					type		: 'TextArea',
					title		: 'Description',
					validators	: []
				},
                policyLabels : {
                    type	   : 'Select2Remote',
                    title	   : 'Policy Label',
                    pluginAttr : this.getPluginAttr(),
                    options    : function(callback, editor){
                        callback();
                    },
                },
				isEnabled : {
					type		: 'Switch',
					title		: '',//localization.tt("lbl.policyStatus"),
					editorClass : 'policy-enable-disable',
					onText		: 'Enabled',
					offText		: 'Disabled',
					width		: '80',
					switchOn	: true
				},
                policyPriority : {
                    type            : 'Switch',
                    title           : '',//localization.tt("lbl.policyStatus"),
                    editorClass     : 'policy-priority',
                    onText          : 'Override',
                    offText         : 'Normal',
                    width           : '80',
                    switchOn        : false
                },
				isAuditEnabled : {
					type		: 'Switch',
					title		: localization.tt("lbl.auditLogging"),
					onText 		: 'Yes',
					offText		: 'No',
					switchOn	: true
				},
				//recursive(ON/OFF) toggle
				recursive : {
					type : 'Switch',
					title : 'Recursive',
					onText : 'On',
					offText : 'Off',
					switchOn : true
				}
			});
		},

		/** need to pass eventTime in queryParams(opt.data) */
		fetchByEventTime : function(opt){
			var queryParams = opt.data;
			queryParams.policyId = this.get('id');
			queryParams.versionNo = this.get('version');
			if(_.isUndefined(queryParams.eventTime)){
				throw('eventTime can not be undefined');
			}

			opt.url = 'service/plugins/policies/eventTime';
			return this.fetch(opt);
		},

		fetchByVersion : function(versionNo, opt){
			if(_.isUndefined(versionNo)){
				throw('versionNo can not be undefined');
			}
			opt.url = 'service/plugins/policy/'+this.get('id')+'/version/'+versionNo;
			return this.fetch(opt);
		},

		fetchVersions : function(){
			var versionList;
			var url = 'service/plugins/policy/'+this.get('id')+'/versionList';
			$.ajax({
				url : url,
				async : false,
				dataType : 'JSON',
				success : function(data){
                                        versionList = (data.value.split(',').map(Number)).sort(function(a, b) { return a - b });
				},
			});
			return versionList;
		},

		/** This models toString() */
		toString : function(){
			return this.get('name');
                },

        getPluginAttr : function(){
            var tags = [];
            _.each(this.get('policyLabels') , function(name){
                tags.push( { 'id' : _.escape( name ), 'text' : _.escape( name ) } );
            });
            return{
                multiple: true,
                closeOnSelect : true,
                placeholder : 'Policy Label',
                allowClear: true,
                tokenSeparators: ["," , " "],
                tags : true,
                initSelection : function (element, callback) {
                    callback(tags);
                },
                createSearchChoice: function(term, data) {
                    term = _.escape(term);
                    if ($(data).filter(function() {
                        return this.text.localeCompare(term) === 0;
                    }).length === 0) {
                        if($.inArray(term, this.val()) >= 0){
                            return null;
                        }else{
                            return {
                                id : term,
                                text: term
                            };
                        }
                    }
                },
                ajax: {
                    url: 'service/plugins/policyLabels',
                    dataType: 'json',
                    data: function (term, page) {
                        return {policyLabel : term};
                    },
                    results: function (data, page) {
                        var results = [] , selectedVals = [];
                        if(data.resultSize != "0"){
                            results = data.map(function(m){	return {id : _.escape(m), text: _.escape(m) };	});
                        }
                        return {results : results};
                    },
                },
                formatResult : function(result){
                    return result.text;
                },
                formatSelection : function(result){
                    return result.text;
                },
            }
        },
    }, {
		// static class members
	});

    return RangerPolicy;
	
});

File: govern/data-security/ranger/intg/src/main/python/apache_ranger/model/ranger_policy.py
#!/usr/bin/env python

#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from apache_ranger.model.ranger_base import *
from apache_ranger.utils             import *


class RangerPolicy(RangerBaseModelObject):
    POLICY_TYPE_ACCESS    = 0
    POLICY_TYPE_DATAMASK  = 1
    POLICY_TYPE_ROWFILTER = 2

    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBaseModelObject.__init__(self, attrs)

        self.service              = attrs.get('service')
        self.name                 = attrs.get('name')
        self.policyType           = attrs.get('policyType')
        self.policyPriority       = attrs.get('policyPriority')
        self.description          = attrs.get('description')
        self.resourceSignature    = attrs.get('resourceSignature')
        self.isAuditEnabled       = attrs.get('isAuditEnabled')
        self.resources            = attrs.get('resources')
        self.additionalResources  = attrs.get('additionalResources')
        self.policyItems          = attrs.get('policyItems')
        self.denyPolicyItems      = attrs.get('denyPolicyItems')
        self.allowExceptions      = attrs.get('allowExceptions')
        self.denyExceptions       = attrs.get('denyExceptions')
        self.dataMaskPolicyItems  = attrs.get('dataMaskPolicyItems')
        self.rowFilterPolicyItems = attrs.get('rowFilterPolicyItems')
        self.serviceType          = attrs.get('serviceType')
        self.options              = attrs.get('options')
        self.validitySchedules    = attrs.get('validitySchedules')
        self.policyLabels         = attrs.get('policyLabels')
        self.zoneName             = attrs.get('zoneName')
        self.conditions           = attrs.get('conditions')
        self.isDenyAllElse        = non_null(attrs.get('isDenyAllElse'), False)

    def type_coerce_attrs(self):
        super(RangerPolicy, self).type_coerce_attrs()

        self.resources            = type_coerce_dict(self.resources, RangerPolicyResource)
        self.additionalResources  = type_coerce_list(self.additionalResources, dict)
        self.policyItems          = type_coerce_list(self.policyItems, RangerPolicyItem)
        self.denyPolicyItems      = type_coerce_list(self.denyPolicyItems, RangerPolicyItem)
        self.allowExceptions      = type_coerce_list(self.allowExceptions, RangerPolicyItem)
        self.denyExceptions       = type_coerce_list(self.denyExceptions, RangerPolicyItem)
        self.dataMaskPolicyItems  = type_coerce_list(self.dataMaskPolicyItems, RangerDataMaskPolicyItem)
        self.rowFilterPolicyItems = type_coerce_list(self.rowFilterPolicyItems, RangerRowFilterPolicyItem)
        self.validitySchedules    = type_coerce_list(self.validitySchedules, RangerValiditySchedule)

        if isinstance(self.additionalResources, list):
            additionalResources = []

            for entry in self.additionalResources:
                additionalResources.append(type_coerce_dict(entry, RangerPolicyResource))

            self.additionalResources = additionalResources
        else:
            self.additionalResources = None

    def add_resource(self, resource):
        if resource is not None:
            if self.resources is None:
                self.resources = resource
            else:
                if self.additionalResources is None:
                    self.additionalResources = []

                self.additionalResources.append(resource)

class RangerPolicyResource(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.values      = attrs.get('values')
        self.isExcludes  = non_null(attrs.get('isExcludes'), False)
        self.isRecursive = non_null(attrs.get('isRecursive'), False)


class RangerPolicyItemCondition(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.type   = attrs.get('type')
        self.values = attrs.get('values')


class RangerPolicyItem(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.accesses      = attrs.get('accesses')
        self.users         = attrs.get('users')
        self.groups        = attrs.get('groups')
        self.roles         = attrs.get('roles')
        self.conditions    = attrs.get('conditions')
        self.delegateAdmin = non_null(attrs.get('delegateAdmin'), False)

    def type_coerce_attrs(self):
        super(RangerPolicyItem, self).type_coerce_attrs()

        self.accesses = type_coerce_list(self.accesses, RangerPolicyItemAccess)


class RangerDataMaskPolicyItem(RangerPolicyItem):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerPolicyItem.__init__(self, attrs)

        self.dataMaskInfo = attrs.get('dataMaskInfo')

    def type_coerce_attrs(self):
        super(RangerDataMaskPolicyItem, self).type_coerce_attrs()

        self.dataMaskInfo = type_coerce(self.dataMaskInfo, RangerPolicyItemDataMaskInfo)


class RangerRowFilterPolicyItem(RangerPolicyItem):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerPolicyItem.__init__(self, attrs)

        self.rowFilterInfo = attrs.get('rowFilterInfo')

    def type_coerce_attrs(self):
        super(RangerRowFilterPolicyItem, self).type_coerce_attrs()

        self.rowFilterInfo = type_coerce(self.rowFilterInfo, RangerPolicyItemRowFilterInfo)


class RangerValiditySchedule(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.startTime = attrs.get('startTime')
        self.endTime   = attrs.get('endTime')
        self.timeZone  = attrs.get('timeZone')


class RangerPolicyItemAccess(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.type      = attrs.get('type')
        self.isAllowed = non_null(attrs.get('isAllowed'), True)


class RangerPolicyItemDataMaskInfo(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.dataMaskType  = attrs.get('dataMaskType')
        self.conditionExpr = attrs.get('conditionExpr')
        self.valueExpr     = attrs.get('valueExpr')


class RangerPolicyItemRowFilterInfo(RangerBase):
    def __init__(self, attrs=None):
        if attrs is None:
            attrs = {}

        RangerBase.__init__(self, attrs)

        self.filterExpr = attrs.get('filterExpr')
